rfcs/framework-0040-relations.md

```diff
@@ -0,0 +1,321 @@
+
+We suggest adding a ``@v`` key in the *dereferenced data* (not in the database) that includes the version counter of the referenced record:
+
+```json
+{"id": "1234", "@v": 18, "a": ""}
+```
+
+It should be investigated if this if sufficient in order to generate proper queries in Elasticsearch. Alternatively, we will need to create a compound key concatenating the id and the version - for instance:
+
+```json
+{"id": "1234", "@v": "v18:1234", "a": ""}
+```
+
+Note, we cannot replace the ``id`` key, as it must be stable between the normalised and denormalised version.
+
+
+### Programmatic API
+
+
+#### Defining relations
+
+Relations could be defined by a new system field named ``RelationsField``. The reason for using the name "relations" is to avoid confusion with REST API HATEOS links in the responses.
+
+```python
+class Record(RecordBase):
+    relations = RelationsField(
+        resource_type=PIDRelation('metadata.resource_type', ResourceTypeRecord.pid),
+        subjects=PKListRelation('metadata.subjects', SubjectsRecord),
+    )
+```
+
+Note that the class property name (``relations``) can be named whatever you like (e.g. ``vocabularies``).
+
+We proposed four relationship types:
+
+- ``PIDRelation`` - a relation using PIDStore for lookup.
+- ``PKRelation`` - a relation using record UUIDs for lookup (i.e. a record's primary key).
+- ``PIDListRelation`` - same as ``PIDRelation`` but for list of referenced objects.
+- ``PKListRelation`` - same as ``PKRelation`` but for list of objects.
+
+An alternative, would be to specify relations individually:
+
+```python
+class Record(RecordBase):
+    resource_type = PIDRelation('metadata.resource_type', ResourceTypeRecord.pid)
+```
+
+However, this will make it a bit more diffcult to operate on all relations at once (e.g. ``record.relations.dereference()``).
+
+##### Limiting attributes
+
+Limiting the attributes could be done as:
+
+```python
+PIDRelation('metadata.resource_type', ResourceTypeRecord.pid, attrs=[
+    'type', 'subtype', 'some.nested.key'])
+```
+
+#### Dereferencing / retrieving the record
+
+```python
+# No relation defined
+record.relations.resource_type() is None
+
+# Getting via metadata
+record['metadata']['resource_type']
+# - return only id if  not dereferenced
+# - return id, type and subtype
+
+# Get the related record:
+# - if not dereferenced: query db and store dereferenced
+# - if dereferenced: use dereferenced
+rt = record.relations.resource_type()
+# Get the related record (force a DB query)
+rt = record.relations.resource_type(force=True)
+
+# Get many records:
+for s in record.relations.subjects():
+    # ...
+
+# Dereference all records at once
+record.relations.dereference()
+```
+
+#### Setting related records
+
+
+```python
+# Via metadata
+Record({
+    "metadata": {
+        "resource_type": {"id": "text-article"}
+    }
+})
+
+# Via system field
+record.relations.resource_type = ResourceType.get_record('text-article')
```

Comment:
Is the difference between this one and the one below:
- this one sets a dereferenced record
- the one below only sets a non-dereferenced record

Otherwise, when would one use this one versus the one below?